File: 004/code.go
// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

package main

import "fmt"
import "strconv"

func main() {
	numbers := threeDigitProducts()
	fmt.Println(max(palindromes(numbers)))
}

func max(numbers []int) (result int) {
	result = numbers[0]
	for _, number := range numbers {
		if number > result {
			result = number
		}
	}
	return
}

func palindromes(numbers []int) []int {
	var result []int
	for _, number := range numbers {
		if isPalindrome(number) == true {
			result = append(result, number)
		}
	}
	return result
}

func isPalindrome(number int) bool {
	s := strconv.Itoa(number)
	return s == Reverse(s)
}

func Reverse(s string) (result string) {
	for _, v := range s {
		result = string(v) + result
	}
	return
}

func threeDigitProducts() []int {
	x := 100
	y := 100
	var result []int
	for x < 1000 {
		y = 100
		for y < 1000 {
			result = append(result, x*y)
			y++
		}
		x++
	}
	return result
}

File: 004/code.jl
# A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product of two 3-digit numbers.

function palindromeNumbers(a, b)
  result = []
  for num in collect(a:b)
    s = string(num)
    if s == reverse(s)
      push!(result, s)
    end
  end
  result
end

function isPalindrome(num)
  s = string(num)
  s == reverse(s)
end

function threeDigitProducts()
  result = []
  for i in collect(100:999)
    for j in collect(i:999)
      push!(result, i*j)
    end
  end
  result
end

nums = sort(threeDigitProducts())
index = findlast(x->isPalindrome(x), nums)
println(nums[index])
